lib/portlayer/storage/nfs/volume.go

```diff
@@ -0,0 +1,260 @@
+	if err != nil {
+		return nil, err
+	}
+
+	v := &NFSv3VolumeStore{
+		Name:           storeName,
+		volumeStoreDir: nfsTargetURL.Path,
+		MH:             mount,
+		Target:         nfsTargetURL,
+		SelfLink:       selfLink,
+	}
+
+	return v, nil
+}
+
+// Returns the path to the vol relative to the given store.  The dir structure
+// for a vol in a nfs store is `<configured nfs server path>/volumes/<vol ID>/<volume contents>`.
+func (v *NFSv3VolumeStore) volDirPath(ID string) string {
+	return path.Join(v.volumeStoreDir, VolumesDir, ID)
+}
+
+func (v *NFSv3VolumeStore) volumesDir() string {
+	return path.Join(v.volumeStoreDir, VolumesDir)
+}
+
+// Returns the path to the metadata directory for a volume
+func (v *NFSv3VolumeStore) volMetadataDirPath(ID string) string {
+	return path.Join(v.volumeStoreDir, metadataDir, ID)
+}
+
+// Creates a volume directory and volume object for NFS based volumes
+func (v *NFSv3VolumeStore) VolumeCreate(op trace.Operation, ID string, store *url.URL, capacityKB uint64, info map[string][]byte) (*storage.Volume, error) {
+	target, err := v.MH.Mount(v.Target)
+	if err != nil {
+		return nil, err
+	}
+
+	defer v.MH.Unmount(target)
+
+	if _, err := target.MkDir(v.volDirPath(ID), filePermissions); err != nil {
+		return nil, err
+	}
+
+	backing := NewNFSVolumeDevice(v.Target, v.volDirPath(ID))
+
+	vol, err := storage.NewVolume(v.SelfLink, ID, info, backing)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := writeMetadata(op, v.volMetadataDirPath(ID), info, target); err != nil {
+		return nil, err
+	}
+
+	op.Infof("nfs volume (%s) successfully created on volume store (%s)", ID, v.Name)
+	return vol, nil
+}
+
+// Removes a volume and all of it's contents from the nfs store. We already know via the cache if it is in use.
+func (v *NFSv3VolumeStore) VolumeDestroy(op trace.Operation, vol *storage.Volume) error {
+	target, err := v.MH.Mount(v.Target)
+	if err != nil {
+		return err
+	}
+
+	defer v.MH.Unmount(target)
+
+	op.Infof("Attempting to remove volume (%s) and it's metadata from volume store (%s)", vol.ID, v.Name)
+
+	//remove volume directory and children
+	if err := target.RemoveAll(v.volDirPath(vol.ID)); err != nil {
+		op.Errorf("Failed to remove volume (%s) on volume store (%s) due to error (%s)", vol.ID, v.Name, err)
+		return err
+	}
+
+	//XXX: what should we do if we lose the nfs server between volume and metadata deletion?
+
+	//remove volume metadata directory and children
+	if err := target.RemoveAll(v.volMetadataDirPath(vol.ID)); err != nil {
+		op.Errorf("Failed to remove metadata for volume (%s) at path (%q) on volume store (%s)", vol.ID, v.volDirPath(vol.ID), v.Name)
+		//FIXME: Should we bail here? the volume is gone at this point...
+	}
+	op.Infof("Successfully removed volume (%s) from volumestore (%s)", vol.ID, v.Name)
+
+	return nil
+}
+
+func (v *NFSv3VolumeStore) VolumesList(op trace.Operation) ([]*storage.Volume, error) {
+	volumes := make([]*storage.Volume, 0)
```

Comment:
filtered out `.` and `..` and moved the statement down. I thought I had already moved the statement lower, but didn't even realize the `.` and `..` cases. My unit tests did not catch this, so i might need to find a way to make sure that filter is tested as well.